Amplify/Core/Support/AsyncChannel.swift

```diff
@@ -0,0 +1,88 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+actor AsyncChannel<Element: Sendable>: AsyncSequence {
+    struct Iterator: AsyncIteratorProtocol, Sendable {
+        private let channel: AsyncChannel<Element>
+
+        init(_ channel: AsyncChannel<Element>) {
+            self.channel = channel
+        }
+
+        mutating func next() async -> Element? {
+            await channel.next()
+        }
+    }
+
+    enum InternalFailure: Error {
+        case cannotSendAfterTerminated
+    }
+    typealias ChannelContinuation = CheckedContinuation<Element?, Never>
+
+    private var continuations: [ChannelContinuation] = []
+    private var elements: [Element] = []
+    private var terminated: Bool = false
+
+    private var hasNext: Bool {
+        !continuations.isEmpty && !elements.isEmpty
+    }
+
+    private var canTerminate: Bool {
+        terminated && elements.isEmpty && !continuations.isEmpty
+    }
+
+    init() {
+    }
+
+    nonisolated func makeAsyncIterator() -> Iterator {
```

Comment:
Is this okay? Will the channel work if we iterate in two different task?

Reply:
this function cannot be isolated in the actor. it also does not mutate any state. It creates the iterator which is used by the sequence and all those functions are isolated within the actor.